### PR TITLE
Implement mixed precision for training and inference

### PR DESCRIPTION
**Pull request by LorenzLamm**
Mixed precision for both training and inference.

Using mixed precision during training allows for larger batch sizes and slightly faster training speed (time-wise bottleneck is still data augmentation, which is not affected here).

For segmentation, there is also a slight speed-up, but the main advantage is that inference requires less GPU memory. Thus, it should now be easily possible to perform the prediction using an 8GB GPU -- previously, 8GB was just about the margin causing problems for some users.

### CODE REVIEW

File: src/membrain_seg/segmentation/training/optim_utils.py [hunk 1]
```diff
@@ -79,14 +82,18 @@ def forward(self, data: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
         data = sigmoid(data)
         mask = target != self.ignore_label
 
         # Compute the cross entropy loss while ignoring the ignore_label
         target_comp = target.clone()
         target_comp[target == self.ignore_label] = 0
         target_tensor = torch.tensor(target_comp, dtype=data.dtype, device=data.device)
-        bce_loss = binary_cross_entropy(data, target_tensor, reduction="none")
+
+        bce_loss = binary_cross_entropy_with_logits(
```
Comment (kevinyamauchi) [inline src/membrain_seg/segmentation/training/optim_utils.py:94]: If I understand correctly, this function applies a sigmoid to the prediction before computing the loss. Have you checked how this impacts training performance?

https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html#torch.nn.BCEWithLogitsLoss

**Review by kevinyamauchi — APPROVED** [review]
LGTM @LorenzLamm ! I have a question below, but it's more of a curiosity.

File: src/membrain_seg/segmentation/training/optim_utils.py (hunk 1 above)
Comment (LorenzLamm) [reply src/membrain_seg/segmentation/training/optim_utils.py:94]: Haven't checked in detail how it influences training performance, except that it re-ran training to see if loss curves behave the same: In this sense, there is no difference.

Including the sigmoid in the loss function is supposed to be more numerically stable, as the optimization uses the log sum exp trick to compute the loss: https://gregorygundersen.com/blog/2020/02/09/log-sum-exp/

I guess this helps especially in case of exploding or vanishing gradients, which I didn't observe during training.

Other than that, the loss function is the same as the one I used previously. Note that the `binary_cross_entropy_with_logits` function uses as input the `orig_data` variable instead of previously `data` where sigmoid is applied above.
So both the previous version and the new version first do sigmoid, and then cross entropy. Only the new version computes gradients in a single pass for more stability.
